library/alloc/src/vec/drain.rs

```diff
@@ -149,7 +166,24 @@ impl<T, A: Allocator> ExactSizeIterator for Drain<'_, T, A> {
 }
 
 #[unstable(feature = "trusted_len", issue = "37572")]
+// SAFETY: `Drain` simply forwards to the underlying slice iterator, which implements `TrustedLen`
+// so the required properties are all preserved.
 unsafe impl<T, A: Allocator> TrustedLen for Drain<'_, T, A> {}
 
+#[doc(hidden)]
+#[unstable(feature = "trusted_random_access", issue = "none")]
+// SAFETY: `Drain` forwards to the underlying slice iterator, which implements `TrustedRandomAccessNoCoerce`,
+// and then reads the items instead of just returning a reference. As `TrustedRandomAccessNoCoerce`
+// requires each index to be accessed only once, this is safe to do here.
+//
+// TrustedRandomAccess (without NoCoerce) must not be implemented because
+// subtypes/supertypes of `T` might not be `NonDrop`
```

Comment:
`Drain` is a bit more complicated than `IntoIter` due to its `Drop` impl doing actual work even for `T: Copy`. This comment should mention why this is still ok. Something along the lines that the tail move does not depend on how much has been drained.